### DISCUSSION TITLE
How trade conditions work (e.g. Trade_WPR)?

### DISCUSSION THREAD

**Discussion by soumsena** [EA31337/EA31337#308] [Q&A] [answered]
In the  function Trade_WPR at line 2962
result = wpr[period][CURR] > **50** + signal_level;
i would like if it is **+50** or **-50**, i don't understand the condition. I usually work with the -50, -80 et -20.

I tried to write this condition:
```
if (iWPR(NULL,piwpr,piwprbar,1)<-80 && iWPR(NULL,piwpr,piwprbar,0)>=-80)
if (METHOD(signal_method, 6)) result &= wpr[period][PREV]  < -80;
if (METHOD(signal_method, 7)) result &= wpr[period][CURR]  >= -80;
```

**Comment by kenorb** [comment, marked as answer]
### Buy

```
    case ORDER_TYPE_BUY:                              
      result = wpr[period][CURR] > 50 + signal_level; 
```

### Sell
```
    case ORDER_TYPE_SELL:                                
      result = wpr[period][CURR] < 50 - signal_level;    
```

You control the signal level with `WPR_SignalLevel` param, so you don't have to change any code.

So for example, if you've `WPR_SignalLevel` set to `30`, then it is:
- `wpr[period][CURR] > 50 + 30` => `-WPR > 80` => `WPR > -80` for buy
- `wpr[period][CURR] < 50 - 30` => `-WPR < 20` => `WPR < -20` for sell

The value is converted from negative to positive at the following line (note the `-`):
```
wpr[index][i] = -iWPR(symbol, tf, (int) (WPR_Period * ratio), i + WPR_Shift);
```

So all you need is to optimize `WPR_SignalLevel` param to find its best value, without changing the code.

Let me know if that make sense.

### Bitwise operators

These `&=` are basically extra conditions on top of the main one.

They are controlled using `(Strategy name)(Timeframe)_SignalMethod` params.

----

Here is the main condition (as mentioned before):
```
    case ORDER_TYPE_BUY:                              
      result = wpr[period][CURR] > 50 + signal_level;
```
Now, `result` is bool type, so it got `true` on condition/signal, or `false` when condition is matched.

Now, `&=` is bitwise AND operation, so:

```
result = true;
result = result & true; // Same as result &= true;
result &= true;
```
both `true` and `true`, makes it `true`.

More examples:
- `true & true & true` => `true`
- `true & true & false & true` => `false`

And this example:
```
result = true;
result = result & true;
result &= true; // Same as above.
result &= false;
```
results in `false`, because one condition is false.

Also note, that `&` (AND) operator is opposite of `|` (OR).

So basically all conditions (using AND bitwise operator) needs to be true, otherwise, if one condition is false, the end result is also false.

So the whole idea is like:
```
result = true;
if (METHOD_1) result &= EXTRA_CONDITION_1;
if (METHOD_2) result &= EXTRA_CONDITION_2;
if (METHOD_3) result &= EXTRA_CONDITION_3;
if (METHOD_4) result &= EXTRA_CONDITION_4;
```
where you can have main condition, and extra condition if needed, which aims to filter out the main signal.

So if you've main condition of `WPR > -80` for buy, and `* WPR < -20` for sell, sometimes it's not good enough, as you want to make your condition more advanced. So you can add extra level of conditions on top of it. E.g.
```
    case ORDER_TYPE_BUY:                              
        result = WPR > -80;
        result &= RSI > 50
```
In this case, you got signal on buy on `WPR > -80`, only when `RSI > 50`, otherwise it won't happen.

----

So basically:
```
result = wpr[period][CURR] > 50 + signal_level;
if (METHOD(signal_method, 0)) result &= wpr[period][CURR] < wpr[period][PREV];
if (METHOD(signal_method, 1)) result &= wpr[period][PREV] < wpr[period][FAR];
if (METHOD(signal_method, 2)) result &= wpr[period][PREV] > 50 + signal_level;
...
```
consist main condition and extra conditions which are controled by `WPR*_SignalMethod` params.

Where `METHOD()` is basically a macro defined in `Condition.mqh`:
```
#define METHOD(method, no) ((method & (1<<no)) == 1<<no)
```

So:
- If `WPR*_SignalMethod` is set to `0`, then signal on the main condition only.
- When `WPR*_SignalMethod` is `1`, then signal when main condition and 1st extra condition (`METHOD(signal_method, 0)` are both true.
- When `WPR*_SignalMethod` is `2`, then signal when main condition and 2nd extra condition is true (`METHOD(signal_method, 1)`).

And here is where it's more confusing:
- When `WPR*_SignalMethod` is `3`, then signal when main condition and 1st and 2nd extra conditions combined are true, because 3 consist 1 and 2 (1+2=3).

So in simple words. signal method param corresponds to the following conditions:
- 0 - means the main condition to signal when `true`
- 1 - main and 1st conditions (`1 & (1<<0)`)
- 2 - main and 2nd conditions (`2 & (1<<1)`)
- 3 - main and 1st and 2nd conditions
- 4 - main and 3rd conditions  (`4 & (1<<2)`)
- 5 - main and 1st and 3rd conditions
- 6 - main, 2nd and 3rd (without 1st)
- 7 - all above conditions need to be true
- 8 - main and 4th conditions
- 15 - all above conditions need to be true
- 16 - main and 5th conditions
- 31 - all above conditions need to be true
- 32 - main and 6th conditions
- 63 - all above conditions need to be true
- 64 - main and 7th conditions
- 127 - all above conditions need to be true
- 128 - main and 8th conditions
- 255 - main and **all** above conditions (from 1st to 8th)
- and so on.

All values between are basically a combination of multiple conditions based on its bit configuration, so you can optimize signal method param with all permutations according to your needs, in order to find its best configuration.

Check [Open methods](https://github.com/EA31337/EA31337/wiki/Open-methods) for further info.
